src/compiler/scala/tools/nsc/transform/Mixin.scala

```diff
@@ -243,11 +243,17 @@ abstract class Mixin extends InfoTransform with ast.TreeDSL {
           case NoSymbol =>
             val isMemberOfClazz = clazz.info.findMember(member.name, 0, 0L, stableOnly = false).alternatives.contains(member)
             if (isMemberOfClazz) {
-              def genForwarder(): Unit = {
-                cloneAndAddMixinMember(mixinClass, member).asInstanceOf[TermSymbol] setAlias member
+              def genForwarder(required: Boolean): Unit = {
+                val owner = member.owner
+                if (owner.isJavaDefined && owner.isInterface && !clazz.parentSymbols.contains(owner)) {
+                  val text = s"Unable to implement a mixin forwarder for $member in $clazz unless interface ${owner.name} is directly extended by $clazz."
+                  if (required) reporter.error(clazz.pos, text)
+                  else warning(clazz.pos, text)
+                } else
+                  cloneAndAddMixinMember(mixinClass, member).asInstanceOf[TermSymbol] setAlias member
               }
 
-              if (settings.XgenMixinForwarders) genForwarder()
+              if (settings.XgenMixinForwarders) genForwarder(required = false)
```

Comment:
I think this will erroneously report a warning instead of an error if both of the following apply:
1. `-XgenMixinForwarders` is used, *and*
2. the forwarder is required for correctness.

I would expect an error to be raised every time a forwarder is required for correctness, regardless of the value of `settings.XgenMixinForwarders`.